### PR TITLE
Add Queue data structure to library

### PR DESCRIPTION
**Pull request by colescott**

### CODE REVIEW

File: library/queue.lisp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    "Peek at the first item of `q`."
+    (if (empty? q)
+        None
+        (Some (peek-unsafe q))))
+
+  (declare peek-unsafe (Queue :a -> :a))
+  (define (peek-unsafe q)
+    "Peek at the first item of `q` without checking if the queue is empty."
+    (lisp :a (q)
+      (cl:car (queue-internal-elements q))))
```
Comment (stylewarning) [inline library/queue.lisp:156]: this could be a type error, which i guess is par for the course for UNSAFE

**Review by stylewarning — APPROVED** [review]
seems ok to me

File: library/queue.lisp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+   #:items!))
+
+(in-package #:coalton-library/queue)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+(cl:declaim (cl:inline make-queue-internal))
```
Comment (eliaslfox) [inline library/queue.lisp:35]: constructors are inline by default

File: library/queue.lisp [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    "Remove and return the first item of `q`."
+    (if (empty? q)
+        None
+        (Some (pop-unsafe! q))))
+
+  (declare pop-unsafe! (Queue :a -> :a))
+  (define (pop-unsafe! q)
+    "Remove and return the first item of `q` without checking if the queue is empty."
+    (lisp :a (q)
+      (cl:decf (queue-internal-length q))
```
Comment (eliaslfox) [inline library/queue.lisp:142]: Reverse the order of these operations

**Review by eliaslfox — APPROVED** [review]

File: library/queue.lisp (hunk 1 above)
Comment (colescott) [reply library/queue.lisp:155]: It seems better for it to error rather than return invalid data. Will fix

**Comment by eliaslfox** [comment]
Add this to the doc generator and add tests

**Pull request merged**